From 386c13e4857e00ebe9f67163fed1ae344b7bf3ab Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hauke.mehrtens@intel.com>
Date: Thu, 10 Sep 2020 22:00:11 +0200
Subject: [PATCH] PONRTSYS-7999: Activate SyncE master in GPHY

This activates the 2048kHz clock input in the 2.5 GPHY and the SyncE
master output.
---
 drivers/net/phy/lantiq.c | 46 +++++++++++++++++++++++++++++++++++++++++++++-
 1 file changed, 45 insertions(+), 1 deletion(-)

diff --git a/drivers/net/phy/lantiq.c b/drivers/net/phy/lantiq.c
index 45644c970905..5ea749123d74 100644
--- a/drivers/net/phy/lantiq.c
+++ b/drivers/net/phy/lantiq.c
@@ -22,6 +22,11 @@
 #define MII_MMDCTRL		0x0d
 #define MII_MMDDATA		0x0e
 
+#define MII_GCTRL		0x09
+#define MII_GCTRL_MSPT		BIT(10)	/* Master/Slave Port Type */
+#define MII_GCTRL_MS		BIT(11)	/* Master/Slave Config Value */
+#define MII_GCTRL_MSEN		BIT(12)	/* Master/Slave Manual Configuration Enable */
+
 #define MII_VR9_11G_IMASK	0x19	/* interrupt mask */
 #define MII_VR9_11G_ISTAT	0x1a	/* interrupt status */
 #define MII_PHY_FWV		0x1e	/* firmware version */
@@ -52,6 +57,13 @@
 #define P31G_ANEG_MGBT_AN_STAT	0x21
 #define STAT_AB_2G5BT_BIT	BIT(5)
 #define STAT_AB_FR_2G5BT	BIT(3)
+#define MMD_DEVVSPEC1			0x1E
+#define VSPEC1_PM_CTRL			0x0C
+#define VSPEC1_PM_CTRL_SYNCE_EN		BIT(4)
+#define VSPEC1_PM_CTRL_SYNCE_CLK_MASK	GENMASK(5, 6)
+#define VSPEC1_PM_CTRL_SYNCE_CLK_PSTN	(0 << 5) /* 8kHz clock */
+#define VSPEC1_PM_CTRL_SYNCE_CLK_EEC1	(1 << 5) /* 2048kHz clock */
+#define VSPEC1_PM_CTRL_SYNCE_CLK_EEC2	(2 << 5) /* 1544kHz clock */
 
 static __maybe_unused int vr9_gphy_mmd_read(struct phy_device *phydev,
 						u16 regnum)
@@ -182,8 +194,25 @@ static __maybe_unused int p31g_gphy_mmd_write(struct phy_device *phydev,
 	return phy_write(phydev, regnum, val);
 }
 
+static __maybe_unused int p31g_gphy_mmd_mask(struct phy_device *phydev,
+					     u32 devad, u32 regnum,
+					     u32 clear, u32 set)
+{
+	int val;
+
+	val = p31g_gphy_mmd_read(phydev, devad, regnum);
+	if (val < 0)
+		return val;
+
+	val &= ~(clear);
+	val |= set;
+	return p31g_gphy_mmd_write(phydev, devad, regnum, val);
+}
+
 static int p31g_gphy_config_init(struct phy_device *phydev)
 {
+	int err, val;
+
 	vr9_gphy_config_init(phydev);
 
 	/* Linux PHY framework does not have a way yet to define
@@ -197,7 +226,22 @@ static int p31g_gphy_config_init(struct phy_device *phydev)
 	dev_info(&phydev->mdio.dev, "fw version 0x%04X\n",
 		 phy_read(phydev, MII_PHY_FWV));
 
-	return 0;
+	/* PRX300 only supports 2.048 MHz input clock, configure that */
+	err = p31g_gphy_mmd_mask(phydev, MMD_DEVVSPEC1, VSPEC1_PM_CTRL,
+				 VSPEC1_PM_CTRL_SYNCE_EN |
+				 VSPEC1_PM_CTRL_SYNCE_CLK_MASK,
+				 VSPEC1_PM_CTRL_SYNCE_EN |
+				 VSPEC1_PM_CTRL_SYNCE_CLK_EEC1);
+	if (err)
+		return err;
+	val = phy_read(phydev, MII_GCTRL);
+	if (val < 0)
+		return val;
+	val |= MII_GCTRL_MS;
+	val |= MII_GCTRL_MSEN;
+	err = phy_write(phydev, MII_GCTRL, val);
+
+	return err;
 }
 
 static int p31g_gphy_config_aneg(struct phy_device *phydev)
